lib/fizzy/instantiate.hpp

```diff
@@ -22,9 +22,42 @@ namespace fizzy
 struct ExecutionResult;
 struct Instance;
 
+using ExecuteFunctionPtr = ExecutionResult (*)(void* context, Instance&, const Value*, int depth);
+using ExecuteFunctionExtendedPtr = ExecutionResult (*)(
+    void* context1, void* context2, Instance&, const Value*, int depth);
+
+struct ExecuteFunction
+{
+    ExecuteFunction() = delete;
+
+    static ExecutionResult func_single_context(
+        void* context1, void* context2, Instance& instance, const Value* args, int depth);
+
+    ExecuteFunction(ExecuteFunctionExtendedPtr f, void* context1, void* context2)
+      : host_func{f, context1, context2}, m_is_host_func{true}
+    {}
+
+    ExecuteFunction(ExecuteFunctionPtr f, void* context = nullptr)
+      : host_func{func_single_context, reinterpret_cast<void*>(f), context}, m_is_host_func{true}
+    {}
+
+    ExecuteFunction(Instance& instance, FuncIdx func_idx)
+      : wasm_func{&instance, func_idx}, m_is_host_func(false)
+    {}
+
+    ExecutionResult operator()(Instance& instance, const Value* args, int depth);
+
+    union
+    {
+        std::tuple<ExecuteFunctionExtendedPtr, void*, void*> host_func;
+        std::tuple<Instance*, FuncIdx> wasm_func;
+    };
+    bool m_is_host_func = false;
```

Comment:
This perhaps better could be

```suggestion
    union
    {
        std::tuple<ExecuteFunctionPtr, void*> host_func; // C++ host func
        std::tuple<ExecuteFunctionExtendedPtr, void*, void*> c_host_func; // C host func
        std::tuple<Instance*, FuncIdx> wasm_func;
    };
    ExecuteFunctionKind kind;
```